src/shell/interpreter.zig

```diff
@@ -602,19 +602,234 @@ pub const EnvMap = struct {
+        if (this.cwd) |*cwd| cwd.deref();
+        for (this.jsobjs.items) |jsobj| {
+            jsobj.unprotect();
+        }
+        if (this.args) |a| a.deinit();
+        bun.destroy(this);
+    }
+
+    pub fn setCwd(this: *ParsedShellScript, globalThis: *JSGlobalObject, callframe: *JSC.CallFrame) callconv(.C) JSC.JSValue {
+        const arguments_ = callframe.arguments(2);
+        var arguments = JSC.Node.ArgumentsSlice.init(globalThis.bunVM(), arguments_.slice());
+        const str_js = arguments.nextEat() orelse {
+            globalThis.throw("$`...`.cwd(): expected a string argument", .{});
+            return .undefined;
+        };
+        const str = bun.String.fromJS(str_js, globalThis);
+        this.cwd = str;
+        return .undefined;
+    }
+
+    pub fn setQuiet(this: *ParsedShellScript, _: *JSGlobalObject, _: *JSC.CallFrame) callconv(.C) JSC.JSValue {
+        log("Interpreter(0x{x}) setQuiet()", .{@intFromPtr(this)});
+        this.quiet = true;
+        return .undefined;
+    }
+
+    pub fn setEnv(this: *ParsedShellScript, globalThis: *JSGlobalObject, callframe: *JSC.CallFrame) callconv(.C) JSC.JSValue {
+        var env =
+            if (this.export_env) |*env|
+        brk: {
+            env.clearRetainingCapacity();
+            break :brk env.*;
+        } else EnvMap.init(bun.default_allocator);
+        defer this.export_env = env;
+
+        const value1 = callframe.argument(0);
+        if (!value1.isObject()) {
+            globalThis.throwInvalidArguments("env must be an object", .{});
+            return .undefined;
+        }
+
+        var object_iter = JSC.JSPropertyIterator(.{
+            .skip_empty_name = false,
+            .include_value = true,
+        }).init(globalThis, value1);
+        defer object_iter.deinit();
+
+        env.ensureTotalCapacity(object_iter.len);
+
+        // If the env object does not include a $PATH, it must disable path lookup for argv[0]
+        // PATH = "";
+
+        while (object_iter.next()) |key| {
+            const keyslice = key.toOwnedSlice(bun.default_allocator) catch bun.outOfMemory();
+            var value = object_iter.value;
+            if (value == .undefined) continue;
+
+            const value_str = value.getZigString(globalThis);
+            const slice = value_str.toOwnedSlice(bun.default_allocator) catch bun.outOfMemory();
+            const keyref = EnvStr.initRefCounted(keyslice);
+            defer keyref.deref();
+            const valueref = EnvStr.initRefCounted(slice);
+            defer valueref.deref();
+
+            env.insert(keyref, valueref);
+        }
+
+        return .undefined;
+    }
+
+    pub fn createParsedShellScript(
+        globalThis: *JSC.JSGlobalObject,
+        callframe: *JSC.CallFrame,
+    ) callconv(.C) JSValue {
+        var shargs = ShellArgs.init();
+
+        const arguments_ = callframe.arguments(2);
+        var arguments = JSC.Node.ArgumentsSlice.init(globalThis.bunVM(), arguments_.slice());
```

Comment:
This leaks memory on errors.

Let's just not use ArgumentsSlice here - it's unnecessary. And, let's create the shell args after validating the arguments. 

```suggestion
        const arguments = arguments_.slice();
        
       if (arguments.len < 2) {
            globalThis.throwNotEnoughArguments("Bun.$", 2, arguments.len);
            return .zero;
        }
        if (arguments
        const string_args = arguments[0];
        const template_args_jst = arguments[1];
```